File: hibernate/src/main/java/io/github/robertomike/hefesto/constructors/ConstructOrderImplementation.java
package io.github.robertomike.hefesto.constructors;

import io.github.robertomike.hefesto.utils.HibernateUtils;
import jakarta.persistence.criteria.*;
import org.hibernate.QueryException;

import java.util.Map;

public class ConstructOrderImplementation extends ConstructOrder {
    private CriteriaBuilder cb;
    private Root<?> root;
    private Map<String, Join<?, ?>> joins;

    public ConstructOrderImplementation setJoins(Map<String, Join<?, ?>> joins) {
        this.joins = joins;
        return this;
    }

    public void construct(CriteriaBuilder cb, CriteriaQuery<?> cr, Root<?> root) {
        this.cb = cb;
        this.root = root;

        cr.orderBy(items.stream().map(this::constructOrder).toList());
    }

    private Order constructOrder(io.github.robertomike.hefesto.actions.Order value) {
        From<?, ?> from = root;
        var field = value.getField();

        if (field.contains(".") && joins.containsKey(field.split(HibernateUtils.DOT_REGEX)[0])) {
            var splitted = field.split(HibernateUtils.DOT_REGEX);
            from = joins.get(splitted[0]);
            field = splitted[1];
        }


        switch (value.getSort()) {
            case ASC -> {
                return cb.asc(from.get(field));
            }
            case DESC -> {
                return cb.desc(from.get(field));
            }
            default -> throw new QueryException("Unsupported sort: " + value.getSort());
        }
    }
}

File: hibernate-criteria-builder/src/main/kotlin/io/github/robertomike/hefesto/constructors/ConstructOrderImplementation.kt
package io.github.robertomike.hefesto.constructors

import io.github.robertomike.hefesto.utils.HibernateUtils
import jakarta.persistence.criteria.*
import org.hibernate.QueryException

/**
 * Criteria API implementation of ORDER BY clause construction.
 * 
 * Converts ORDER BY definitions into JPA Criteria API Order objects.
 * Supports:
 * - Ascending and descending sort orders
 * - Ordering by fields on the root entity
 * - Ordering by fields on joined entities
 * - Multiple order criteria
 * 
 * Field paths can reference joined entities using dot notation or aliases.
 */
class ConstructOrderImplementation : ConstructOrder() {
    private lateinit var cb: CriteriaBuilder
    private lateinit var root: Root<*>
    private var joins: Map<String, Join<*, *>> = emptyMap()

    /**
     * Sets the join map for resolving order fields on joined entities.
     *
     * @param joins map of join aliases to Join objects
     * @return this instance for chaining
     */
    fun setJoins(joins: Map<String, Join<*, *>>): ConstructOrderImplementation {
        this.joins = joins
        return this
    }

    /**
     * Constructs and applies all ORDER BY clauses to the CriteriaQuery.
     *
     * @param cb the CriteriaBuilder for creating order expressions
     * @param cr the CriteriaQuery to apply ordering to
     * @param root the root entity
     */
    fun construct(cb: CriteriaBuilder, cr: CriteriaQuery<*>, root: Root<*>) {
        this.cb = cb
        this.root = root

        cr.orderBy(items.map { constructOrder(it) })
    }

    private fun constructOrder(value: io.github.robertomike.hefesto.actions.Order): Order {
        var from: From<*, *> = root
        var field = value.field

        if (field.contains(".") && joins.containsKey(field.split(HibernateUtils.DOT_REGEX.toRegex())[0])) {
            val splitted = field.split(HibernateUtils.DOT_REGEX.toRegex())
            from = joins[splitted[0]]!!
            field = splitted[1]
        }

        return when (value.sort) {
            io.github.robertomike.hefesto.enums.Sort.ASC -> cb.asc(from.get<Any>(field))
            io.github.robertomike.hefesto.enums.Sort.DESC -> cb.desc(from.get<Any>(field))
            else -> throw QueryException("Unsupported sort: ${value.sort}")
        }
    }
}
